### PR TITLE
Transition lock to post upgrade statuses on `juju upgrade-series` complete.

### PR DESCRIPTION
**Pull request by ExternalReality**
Complete series hook should fire upon completion command `juju upgrade series complete <machine>`.

### CODE REVIEW

File: api/uniter/unit_test.go [hunk 1]
```diff
@@ -732,7 +732,7 @@ func (s *unitSuite) TestSetUpgradeSeriesStatusUpdatesStatus(c *gc.C) {
 	s.CreateUpgradeSeriesLock(c)
 
 	// Change the state to something other than the default remote state of UnitStarted
-	err := s.apiUnit.SetUpgradeSeriesStatus(arbitaryNonDefaultStatus)
+	err := s.apiUnit.SetUpgradeSeriesStatus(arbitraryNonDefaultStatus)
```
Comment (ExternalReality) [inline api/uniter/unit_test.go]: Drive-by typo fix.

**Comment by ExternalReality** [comment]
[Transition lock states to post upgrade status when complete command is run.](https://trello.com/c/YNhsrT7J/386-transition-lock-states-to-post-upgrade-status-when-complete-command-is-run)

**Comment by ExternalReality** [comment]
[Transition lock states to post upgrade status when complete command is run.](https://trello.com/c/YNhsrT7J/386-transition-lock-states-to-post-upgrade-status-when-complete-command-is-run)

File: state/machine_test.go [hunk 2]
```diff
@@ -2446,6 +2447,27 @@ func (s *MachineSuite) setupTestUpdateMachineSeries(c *gc.C) *state.Machine {
 	return mach
 }
 
+func (s *MachineSuite) addMachineUnit(c *gc.C, mach *state.Machine) *state.Unit {
```
Comment (ExternalReality) [inline state/machine_test.go]: We need to add arbitrary units in order to test the machine functionality related to upgrade series.

File: state/machine.go [hunk 3]
```diff
@@ -2322,6 +2379,21 @@ func (m *Machine) UpdateOperation() *UpdateMachineOperation {
 	return &UpdateMachineOperation{m: &Machine{st: m.st, doc: m.doc}}
 }
 
+func (m *Machine) getUpgradeSeriesLock() (*upgradeSeriesLockDoc, error) {
```
Comment (hmlanigan) [inline state/machine.go]: let's use the new getUpgradeSeriesLock() in machine SetUpgradeSeriesStatus() and UpgradeSeriesStatus() too.

Comment (hmlanigan) [inline state/machine.go]: Can we have a new struct here with the data, rather than the Doc struct, to return here? upgradeSeriesLock? Looking at the pattern set with getInstanceData()

File: apiserver/facades/client/machinemanager/machinemanager.go [hunk 4]
```diff
@@ -452,6 +452,18 @@ func (mm *MachineManagerAPI) updateOneMachineSeries(arg params.UpdateSeriesArg)
...
+		return errors.Trace(err)
+	}
+	machine, err := mm.st.Machine(machineTag.Id())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return machine.CompleteUpgradeSeries()
+}
+
 func (mm *MachineManagerAPI) removeUpgradeSeriesLock(arg params.UpdateSeriesArg) error {
```
[the remark this answers is not in the document]
Comment (ExternalReality) [reply apiserver/facades/client/machinemanager/machinemanager.go]: Still needed just not called yet

File: state/machine.go [hunk 5]
```diff
@@ -2302,17 +2349,27 @@ func setUpgradeSeriesTxnOps(machineDocId, unitName string, unitIndex int, status
...
 		return false, nil
 	}
 	if err != nil {
-		return false, fmt.Errorf("cannot get upgrade series lock for machine %v: %v", m.Id(), err)
+		return false, err
+	}
+	return true, nil
+}
+
+func (m *Machine) isUpgradeSeriesPrepareComplete() (bool, error) {
```
[the remark this answers is not in the document]
Comment (ExternalReality) [reply state/machine.go]: Should we check it here?

File: state/machine.go (hunk 3 above)
Comment (ExternalReality) [reply state/machine.go]: getMachineDoc doesn't do this. it gets the doc directly. Is what getInstanceData() doing a pattern or just a one off?

File: state/machine.go [hunk 6]
```diff
@@ -2302,17 +2349,27 @@ func setUpgradeSeriesTxnOps(machineDocId, unitName string, unitIndex int, status
...
+	lock, err := m.getUpgradeSeriesLock()
+	if err != nil {
+		return false, err
+	}
+	for _, prepareUnitStatus := range lock.PrepareUnits {
```
[the remark this answers is not in the document]
Comment (ExternalReality) [reply state/machine.go]: Sounds good! forgot about that one.

File: state/machine.go (hunk 3 above)
Comment (ExternalReality) [reply state/machine.go]: I think methods like `isUpgradeSeriesComplete` and higher level patterns might be better. In truth I think `isUpgradeSeriesComplete` is too `ask` rather than `tell`. I think we can come up with some nice methods here rather than keep on asking for a lock to examine. We might be able to move methods onto a lock object if nice patterns emerge.

File: apiserver/facades/client/machinemanager/machinemanager.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (hmlanigan) [reply apiserver/facades/client/machinemanager/machinemanager.go]: please comment with a TODO

File: state/machine.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (hmlanigan) [reply state/machine.go]: yes

File: state/machine.go (hunk 3 above)
Comment (hmlanigan) [reply state/machine.go]: Can we add a TODO in SetUpgradeSeriesStatus() and UpgradeSeriesStatus()?

Comment (hmlanigan) [reply state/machine.go]: I haven't checked beyond the state/machine.  A Doc seems only be directly accessed when changing it.  Getting data from a doc is returned in a single type, or a struct not of the Doc type.

File: state/machine.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ExternalReality) [reply state/machine.go]: I disagree, I wrote this method with the intention that it assumes the doc is alive. I'll gladly add a check to the call site.

File: state/machine.go (hunk 3 above)
Comment (ExternalReality) [reply state/machine.go]: I don't see the pattern. I see some methods in the package and other packages which do things like return entire raw collections. I don't see an issue with these since it they are unexported methods. None of the exported methods return the entire document directly.

File: state/machine.go [hunk 7]
```diff
@@ -2247,28 +2247,10 @@ func (m *Machine) SetUpgradeSeriesStatus(unitName string, status model.UnitSerie
 		if err := m.isStillAlive(); err != nil {
 			return nil, errors.Trace(err)
 		}
-		coll, closer := m.st.db().GetCollection(machineUpgradeSeriesLocksC)
-		defer closer()
-		var lock upgradeSeriesLockDoc
-		err := coll.FindId(m.Id()).One(&lock)
+		docIndex, err := m.getUnitIndex(unitName, status)
```
Comment (ExternalReality) [inline state/machine.go]: @hmlanigan what do you think of routines like this which hide the extraction of the lock. Taking the lock as a parameter would alllow us to chain these things together at which point a little builder pattern would start to emerge, but I don't see that just yet since nothing yet calls for more than one of these little util methods. However, I was trying to hold off since the shape of the lock type may be forcing some things that could end up changing entirely. Don't want to jump the gun here.

**Comment by ExternalReality** [comment]
!!build!!

File: state/machine.go [hunk 8]
```diff
@@ -2184,6 +2185,42 @@ func (m *Machine) RemoveUpgradeSeriesLock() error {
 	return nil
 }
 
+func (m *Machine) CompleteUpgradeSeries() error {
```
Comment (hmlanigan) [inline state/machine.go]: Needs header comment

File: state/machine.go [hunk 9]
```diff
@@ -2184,6 +2185,42 @@ func (m *Machine) RemoveUpgradeSeriesLock() error {
...
+			return nil, errors.Trace(err)
+		}
+		if !completed {
+			return nil, fmt.Errorf("machine %q has not finished preparing", m.Id())
+		}
+		lock, err := m.getUpgradeSeriesLock()
+		if err != nil {
+			return nil, err
+		}
+		for _, unit := range lock.CompleteUnits {
```
Comment (hmlanigan) [inline state/machine.go]: the machine complete status should be updated to Started here also

File: state/machine.go [hunk 10]
```diff
@@ -2184,6 +2185,42 @@ func (m *Machine) RemoveUpgradeSeriesLock() error {
...
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		if attempt > 0 {
+			if err := m.Refresh(); err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		if err := m.isStillAlive(); err != nil {
+			return nil, errors.Trace(err)
+		}
+		completed, err := m.isUpgradeSeriesPrepareComplete()
```
Comment (hmlanigan) [inline state/machine.go]: check that the machine complete status is NotStarted too.

File: state/machine.go [hunk 11]
```diff
@@ -2218,40 +2255,90 @@ func (m *Machine) SetUpgradeSeriesStatus(unitName string, status model.UnitSerie
...
+		},
+		{
+			C:      machineUpgradeSeriesLocksC,
+			Id:     machineDocID,
+			Update: bson.D{{"$set", bson.D{{"prepare-status", status}}}},
+		},
+	}
+}
+
+func completeUpgradeSeriesTxnOps(machineDocID string, units []unitStatus) []txn.Op {
```
Comment (hmlanigan) [inline state/machine.go]: the code would be easier to read if this was directly below CompleteUpgradeSeries()

File: state/machine.go (hunk 7 above)
Comment (hmlanigan) [reply state/machine.go]: i like the little helper routines.

File: state/machine.go [hunk 12]
```diff
@@ -2218,40 +2255,90 @@ func (m *Machine) SetUpgradeSeriesStatus(unitName string, status model.UnitSerie
...
 	err := m.st.db().Run(buildTxn)
 	if err != nil {
 		err = onAbort(err, ErrDead)
-		logger.Errorf("cannot set series upgrade status for unit %q of machine %q: %v", unitName, m.Id(), err)
 		return err
 	}
+	return nil
+}
 
+// SetUpgradeSeriesStatus sets the machine status of a series upgrade.
```
Comment (hmlanigan) [inline state/machine.go]: s/the machine status/the machine prepare status/

File: state/machine.go [hunk 13]
```diff
@@ -2218,40 +2255,90 @@ func (m *Machine) SetUpgradeSeriesStatus(unitName string, status model.UnitSerie
...
+	}
+	err := m.st.db().Run(buildTxn)
+	if err != nil {
+		err = onAbort(err, ErrDead)
+		return err
+	}
 	return nil
 }
 
+func (m *Machine) isMachineUpgradeSeriesStatusSet(status model.MachineSeriesUpgradeStatus) (bool, error) {
```
Comment (hmlanigan) [inline state/machine.go]: isMachineUpgradeSeriesPrepareStatusSet

File: apiserver/facades/client/machinemanager/machinemanager.go [hunk 14]
```diff
@@ -404,7 +404,7 @@ func (mm *MachineManagerAPI) UpgradeSeriesComplete(args params.UpdateSeriesArg)
 	if err := mm.check.ChangeAllowed(); err != nil {
 		return params.ErrorResult{}, err
 	}
-	err := mm.removeUpgradeSeriesLock(args)
+	err := mm.completeUpgradeSeries(args)
```
Comment (hmlanigan) [inline apiserver/facades/client/machinemanager/machinemanager.go]: machinemanager_test.go needs tests for UpgradeSeriesComplete.

File: state/machine.go (hunk 13 above)
Comment (ExternalReality) [reply state/machine.go]: Lets wait on this one.

File: state/machine.go (hunk 11 above)
Comment (ExternalReality) [reply state/machine.go]: I would rather keep them grouped.

File: state/machine.go [hunk 15]
```diff
@@ -2317,11 +2405,55 @@ func (m *Machine) IsLocked() (bool, error) {
...
+	}
+	return lock.PrepareStatus == model.MachineSeriesUpgradeComplete, nil
+}
+
 // UpdateOperation returns a model operation that will update the machine.
 func (m *Machine) UpdateOperation() *UpdateMachineOperation {
 	return &UpdateMachineOperation{m: &Machine{st: m.st, doc: m.doc}}
 }
 
+func (m *Machine) getUnitIndex(unitName string, status model.UnitSeriesUpgradeStatus) (int, error) {
```
[the remark this answers is not in the document]
Comment (ExternalReality) [reply state/machine.go]: This is going away really soon.

File: state/machine.go (hunk 3 above)
Comment (ExternalReality) [reply state/machine.go]: Lets wait on this

**Comment by ExternalReality** [comment]
$$MERGE$$
